docs/man/rnpkeys.1
'\" t
.\"     Title: rnpkeys
.\"    Author: RNP
.\" Generator: Asciidoctor 2.0.26
.\"      Date: 2026-09-02
.\"    Manual: RNP Manual
.\"    Source: RNP 0.18.1
.\"  Language: English
.\"
.TH "RNPKEYS" "1" "2026-09-02" "RNP 0.18.1" "RNP Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
rnpkeys \- OpenPGP key management utility.
.SH "SYNOPSIS"
.sp
\fBrnpkeys\fP [\fI\-\-homedir\fP \fIdir\fP] [\fIOPTIONS\fP] \fICOMMAND\fP
.SH "DESCRIPTION"
.sp
The \fIrnpkeys\fP command\-line utility is part of the \fIRNP\fP suite and
provides OpenPGP key management functionality, including:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
key listing;
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
key generation;
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
key import/export; and
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
key editing.
.RE
.SS "BASICS"
.sp
By default, \fBrnp\fP will apply a \fICOMMAND\fP, additionally configured with \fIOPTIONS\fP,
to all \fIINPUT_FILE\fP(s) or \fIstdin\fP if no \fIINPUT_FILE\fP is given.
There are some special cases for \fIINPUT_FILE\fP :
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fI\-\fP (dash) substitutes to \fIstdin\fP
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
env:VARIABLE_NAME substitutes to the contents of environment variable VARIABLE_NAME
.RE
.sp
Depending on the input, output may be written:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
to the specified file with a removed or added file extension (\fI.pgp\fP, \fI.asc\fP, \fI.sig\fP); or
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
to \fIstdout\fP.
.RE
.sp
Without the \fB\-\-armor\fP option, output will be in binary.
.sp
If \fICOMMAND\fP requires public or private keys, \fBrnp\fP will look for the keyrings in \fB~/.rnp\fP. The options \fB\-\-homedir\fP and \fB\-\-keyfile\fP override this (see below).
.sp
If \fICOMMAND\fP needs a password, \fBrnp\fP will ask for it via \fBstdin\fP or \fBtty\fP,
unless the \fB\-\-password\fP or \fB\-\-pass\-fd\fP option was specified.
.sp
By default, \fBrnpkeys\fP will use keyrings stored in the \fI~/.rnp\fP directory.
.sp
This behavior may be overridden with the \fI\-\-homedir\fP option.
.sp
If \fICOMMAND\fP needs a password, the command will prompt the caller
via \fIstdin\fP or \fItty\fP, unless the \fB\-\-password\fP or \fB\-\-pass\-fd\fP
options were also used.
.SS "SPECIFYING KEYS"
.sp
Most \fBrnpkeys\fP commands require a key locator or a filter,
representing one or more keys.
.sp
It may be specified in one of the following ways:
.sp
\fBuserid\fP
.RS 4
Or just part of the \fBuserid\fP.
For \fIalice@rnpgp.com\fP, the following methods are considered identical:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fIalice\fP
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fIalice@rnpgp\fP
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fIrnpgp.com\fP
.RE
.RE
.sp
\fBkeyid\fP
.RS 4
Or its right\-most 8 characters. With or without \fI0x\fP at the beginning and spaces/tabs inside. Such as:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fI0x725F6F2D6D5F6120\fP
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fI"725F6F2D 6D5F6120"\fP
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fI0x6D5F6120\fP
.RE
.RE
.sp
\fBkey fingerprint\fP: The 40\-character key fingerprint, such as:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fI"0x416E746F 6E537669 72696465 6E6B6F20"\fP
.RE
.SH "COMMANDS"
.SS "INFORMATIONAL"
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Displays a short help message. No options are expected.
.RE
.sp
\fB\-V\fP, \fB\-\-version\fP
.RS 4
Displays version information. No options are expected.
.RE
.sp
\fB\-l\fP, \fB\-\-list\-keys\fP
.RS 4
List out keys and some brief information about each.
.br
.sp
Additional options:
.sp
\fB\-\-with\-sigs\fP
.RS 4
Additionally display signatures of listed keys.
.RE
.RE
.SS "KEY GENERATION"
.sp
\fB\-g\fP, \fB\-\-generate\-key\fP
.RS 4
Generate a new keypair.
.br
.sp
Without additional options, an RSA primary key pair with an RSA sub\-key pair will be generated, and prompting for the encryption password afterwards.
.sp
Additional options:
.sp
\fB\-\-numbits\fP
.RS 4
Overrides the default key size. For RSA it is \fB3072\fP bits.
.RE
.sp
\fB\-\-expiration\fP \fITIME\fP
.RS 4
Set key and subkey expiration time, counting from the creation time.
.br
.sp
By default generated keys do not expire.
.br
.sp
Expiration time can be specified as:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
expiration date in the ISO 8601:2019 date format (\fIyyyy\-mm\-dd\fP); or
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
hours/days/months/years since creation time with the syntax of \fI20h\fP/\fI30d\fP/\fI1m\fP/\fI1y\fP;
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
number of seconds.
.RE
.RE
.sp
\fB\-\-expert\fP
.RS 4
Select key algorithms interactively and override default settings.
.RE
.sp
\fB\-\-userid\fP
.RS 4
Specifies the \fIuserid\fP to be used in generation.
.RE
.sp
\fB\-\-hash\fP
.RS 4
Specify the hash algorithm used in generation.
.RE
.sp
\fB\-\-cipher\fP
.RS 4
Specify the encryption algorithm used in generation.
.RE
.sp
\fB\-\-s2k\-iterations\fP
.RS 4
Specify the number of iterations for the S2K (string\-to\-key) process.
.br
.sp
This is used during the derivation of the symmetric key, which
encrypts a secret key from the password.
.br
.RE
.sp
\fB\-\-s2k\-msec\fP
.RS 4
Specify that \fBrnpkeys\fP should automatically pick a
\fB\-\-s2k\-iterations\fP value such that the single key derivation operation
would take \fINUMBER\fP of milliseconds on the current system.
.br
.sp
For example, setting it to \fI2000\fP would mean that each secret key
decryption operation would take around 2 seconds (on the current machine).
.RE
.sp
\fB\-\-allow\-weak\-hash\fP
.RS 4
Allow usage of a weak hash algorithm.
.RE
.sp
\fB\-\-allow\-sha1\-key\-sigs\fP
.RS 4
Allow usage of a SHA\-1 key signatures.
.RE
.RE
.SS "KEY/SIGNATURE IMPORT"
.sp
\fB\-\-import\fP, \fB\-\-import\-keys\fP, \fB\-\-import\-sigs\fP
.RS 4
Import keys or signatures.
.br
.sp
While \fBrnpkeys\fP automatically detects the input data format,
one may still wish to specify whether the input provides keys or signatures.
.br
.sp
By default, the import process will stop on the first discovered
erroneous key or signature.
.br
.sp
Additional options:
.sp
\fB\-\-permissive\fP
.RS 4
Skip errored or unsupported packets during the import process.
.RE
.RE
.SS "KEY/SIGNATURE EXPORT"
.sp
\fB\-\-export\-key\fP [\fB\-\-userid\fP=\fIFILTER\fP] [\fIFILTER\fP]
.RS 4
Export key(s). Only export keys that match \fIFILTER\fP if \fIFILTER\fP is given.
.br
.sp
If filter matches a primary key, the subkeys of the primary key are also exported.
.sp
By default, key data is written to \fIstdout\fP in ASCII\-armored format.
.sp
Additional options:
.sp
\fB\-\-output\fP \fIPATH\fP
.RS 4
Specifies output to be written to a file name instead of \fIstdout\fP.
.RE
.sp
\fB\-\-secret\fP
.RS 4
Without this option specified, the command will only export public key(s).
This option must be provided to export secret key(s).
.RE
.RE
.sp
\fB\-\-export\-rev\fP \fIKEY\fP
.RS 4
Export the revocation signature for a specified secret key.
.br
.sp
The revocation signature can be used later in a case of key loss or compromise.
.sp
Additional options:
.sp
\fB\-\-rev\-type\fP
.RS 4
Specifies type of key revocation.
.RE
.sp
\fB\-\-rev\-reason\fP
.RS 4
Specifies reason for key revocation.
.RE
.RE
.SS "KEY MANIPULATION"
.sp
\fB\-\-revoke\-key\fP \fIKEY\fP
.RS 4
Issue revocation signature for the secret key, and save it in the keyring.
.br
.sp
Revoked keys cannot be used further.
.br
.sp
Additional options:
.sp
\fB\-\-rev\-type\fP
.RS 4
Specifies type of key revocation, see \fBoptions\fP section for the available values.
.RE
.sp
\fB\-\-rev\-reason\fP
.RS 4
Specifies reason for key revocation.
.RE
.RE
.sp
\fB\-\-remove\-key\fP \fIKEY\fP
.RS 4
Remove the specified key.
.br
.sp
If a primary key is specified, then all of its subkeys are also removed.
.br
.sp

CHANGELOG.md
## Changelog

### 0.18.1 [2025-11-21]

#### Security

* Fixed a vulnerability in session key generation for PKESK (Public Key Encrypted Session Key) packets.
Session keys were generated without cryptographically random values.

* CVE: [CVE-2025-13470](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2025-13470)
* Ribose CNA Advisory: https://cve.ribose.com/advisories/ra-2025-11-20/
* Red Hat CVE: https://access.redhat.com/security/cve/cve-2025-13402  

### 0.18.0 [2025-05-24] -- RETRACTED, DON'T USE

**SECURITY WARNING:** This version contains a vulnerability in session key generation for
PKESK (Public Key Encrypted Session Key) packets. Session keys are generated
without cryptographically random values. All users must upgrade
to 0.18.1 or later immediately.

#### General

* Discourage use of EAX AEAD mode
* Generate RSA 3072-bit keys by default
* Support dearmoring of GnuPG-armored files (with ARMORED FILE header)
* Generate rnp_ver.h header
* Support DSA 4096 bit keys as some entities use them
* Mark signatures, produced by encrypt-only key or subkey, as invalid
* Allow extra spaces during armored key import
* Better support of Botan 3.5.0+
* Compatibility fixes for different systems
* Update hash function from the weak one on change of the key expiration
* Do not allow 64-bit ciphers for encryption without explicit option
* Internal refactoring and performance updates

#### FFI

* Added functions rnp_signature_error_count()/rnp_signature_error_at() to check why signature validation failed.
* Added functions to create and customize key certifications: rnp_key_certification_create(), rnp_key_signature_set_*/rnp_key_signature_get_*

### 0.17.1 [2024-04-08]

#### General

* Added support for Botan 3.
* Updated support for OpenSSL 3.
* Added support for mimemode in literal data packet.
* Relaxed Base64 decoding to allow spaces after the checksum.

#### FFI

* Added function `rnp_signature_get_features()`.

### 0.17.0 [2023-05-01]

#### General

* Added support for hidden recipient during decryption.
* Added support for AEAD-OCB for OpenSSL backend.
* Improve support for offline secret keys during default key selection.
* Support for GnuPG 2.3+ secret key store format.
* SExp parsing code is moved to separate library, https://github.com/rnpgp/sexp.
* Mark subkeys as expired instead of invalid if primary key is expired.
* AEAD: use OCB by default instead of EAX.
* Do not attempt to validate signatures of unexpected types.
* Use thread-safe time and date handling functions.
* Added ENABLE_BLOWFISH, ENABLE_CAST5 and ENABLE_RIPEMD160 build time options.
* Do not use `EVP_PKEY_CTX_set_dsa_paramgen_q_bits()` if OpenSSL backend version is < 1.1.1e.
* Corrected usage of CEK/KEK algorithms if those differs.

#### FFI

* Added function `rnp_signature_export()`.
* Added flag `RNP_VERIFY_ALLOW_HIDDEN_RECIPIENT` to `rnp_op_verify_set_flags()`.

#### CLI

* Added default armor message type for `--enarmor` command.
* Added command `--set-filename` to specify which file name should be stored in message.
* Added `--add-subkey` subcommand to the `--edit-key`.
* Added `set-expire` subcommand to the `--edit-key`.
* Added `--s2k-iterations` and `--s2k-msec` options to the `rnp`.
* Added `--allow-weak-hash` command to allow usage of weak hash algorithms.
* Report number of new/updated keys during the key import.

### 0.16.3 [2023-04-11]

#### Security

* Fixed issue with possible hang on malformed inputs (CVE-2023-29479).
* Fixed issue where in some cases, secret keys remain unlocked after use (CVE-2023-29480).

### 0.16.2 [2022-09-20]

#### General

* Fixed CMake issues with ENABLE_IDEA and ENABLE_BRAINPOOL.

### 0.16.1 [2022-09-06]

#### General

* Ensure support for RHEL9/CentOS Stream 9/Fedora 36, updating OpenSSL backend support for v3.0.
* Optional import and export of base64-encoded keys.
* Optional raw encryption of the data.
* Optional overriding of the current timestamp.
* Do not fail completely on unknown signature versions.
* Do not fail completely on unknown PKESK/SKESK packet versions.
* Support armored messages without empty line after the headers.
* Added automatic feature detection based on backend.

#### Security

* Separate security rules for the data and key signatures, extending SHA1 key signature support till the Jan, 19 2024.
* Set default key expiration time to 2 years.
* Limit maximum AEAD chunk bits to 16.

src/rnpkeys/rnpkeys.c
/*
 * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
 * Copyright (c) 2009 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is originally derived from software contributed to
 * The NetBSD Foundation by Alistair Crooks (agc@netbsd.org), and
 * carried further by Ribose Inc (https://www.ribose.com).
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
/* Command line program to perform rnp operations */

#include <getopt.h>
#include <regex.h>
#include <string.h>
#include <rnp/rnp.h>
#include "crypto.h"
#include <rnp/rnp_def.h>
#include "../rnp/rnpcfg.h"
#include "rnpkeys.h"
#include <librepgp/stream-common.h>
#include "utils.h"

extern char *__progname;

const char *usage = "--help OR\n"
                    "\t--export-key [options] OR\n"
                    "\t--find-key [options] OR\n"
                    "\t--generate-key [options] OR\n"
                    "\t--import-key [options] OR\n"
                    "\t--list-keys [options] OR\n"
                    "\t--list-sigs [options] OR\n"
                    "\t--trusted-keys [options] OR\n"
                    "\t--get-key keyid [options] OR\n"
                    "\t--version\n"
                    "where options are:\n"
                    "\t[--cipher=<cipher name>] AND/OR\n"
                    "\t[--coredumps] AND/OR\n"
                    "\t[--expert] AND/OR\n"
                    "\t[--force] AND/OR\n"
                    "\t[--hash=<hash alg>] AND/OR\n"
                    "\t[--homedir=<homedir>] AND/OR\n"
                    "\t[--keyring=<keyring>] AND/OR\n"
                    "\t[--output=file] file OR\n"
                    "\t[--keystore-format=<format>] AND/OR\n"
                    "\t[--userid=<userid>] AND/OR\n"
                    "\t[--verbose]\n";

struct option options[] = {
  /* key-management commands */
  {"list-keys", no_argument, NULL, CMD_LIST_KEYS},
  {"list-sigs", no_argument, NULL, CMD_LIST_SIGS},
  {"find-key", optional_argument, NULL, CMD_FIND_KEY},
  {"export", no_argument, NULL, CMD_EXPORT_KEY},
  {"export-key", optional_argument, NULL, CMD_EXPORT_KEY},
  {"import", no_argument, NULL, CMD_IMPORT_KEY},
  {"import-key", no_argument, NULL, CMD_IMPORT_KEY},
  {"gen", optional_argument, NULL, CMD_GENERATE_KEY},
  {"gen-key", optional_argument, NULL, CMD_GENERATE_KEY},
  {"generate", optional_argument, NULL, CMD_GENERATE_KEY},
  {"generate-key", optional_argument, NULL, CMD_GENERATE_KEY},
  {"get-key", no_argument, NULL, CMD_GET_KEY},
  {"trusted-keys", optional_argument, NULL, CMD_TRUSTED_KEYS},
  {"trusted", optional_argument, NULL, CMD_TRUSTED_KEYS},
  /* debugging commands */
  {"help", no_argument, NULL, CMD_HELP},
  {"version", no_argument, NULL, CMD_VERSION},
  {"debug", required_argument, NULL, OPT_DEBUG},
  /* options */
  {"coredumps", no_argument, NULL, OPT_COREDUMPS},
  {"keyring", required_argument, NULL, OPT_KEYRING},
  {"keystore-format", required_argument, NULL, OPT_KEY_STORE_FORMAT},
  {"userid", required_argument, NULL, OPT_USERID},
  {"format", required_argument, NULL, OPT_FORMAT},
  {"hash-alg", required_argument, NULL, OPT_HASH_ALG},
  {"hash", required_argument, NULL, OPT_HASH_ALG},
  {"algorithm", required_argument, NULL, OPT_HASH_ALG},
  {"home", required_argument, NULL, OPT_HOMEDIR},
  {"homedir", required_argument, NULL, OPT_HOMEDIR},
  {"numbits", required_argument, NULL, OPT_NUMBITS},
  {"ssh", no_argument, NULL, OPT_SSHKEYS},
  {"ssh-keys", no_argument, NULL, OPT_SSHKEYS},
  {"sshkeyfile", required_argument, NULL, OPT_SSHKEYFILE},
  {"verbose", no_argument, NULL, OPT_VERBOSE},
  {"pass-fd", required_argument, NULL, OPT_PASSWDFD},
  {"results", required_argument, NULL, OPT_RESULTS},
  {"cipher", required_argument, NULL, OPT_CIPHER},
  {"expert", no_argument, NULL, OPT_EXPERT},
  {"output", required_argument, NULL, OPT_OUTPUT},
  {"force", no_argument, NULL, OPT_FORCE},
  {"secret", no_argument, NULL, OPT_SECRET},
  {NULL, 0, NULL, 0},
};

/* match keys, decoding from json if we do find any */
static int
match_keys(rnp_cfg_t *cfg, rnp_t *rnp, FILE *fp, char *f, const int psigs)
{
    char *json = NULL;
    int   idc;

    if (f == NULL) {
        if (!rnp_list_keys_json(rnp, &json, psigs)) {
            return 0;
        }
    } else {
        if (rnp_match_keys_json(rnp, &json, f, rnp_cfg_getstr(cfg, CFG_KEYFORMAT), psigs) ==
            0) {
            return 0;
        }
    }
    idc = rnp_format_json(fp, json, psigs);
    /* clean up */
    free(json);
    return idc;
}

void
print_praise(void)
{
    (void) fprintf(stderr,
                   "%s\nAll bug reports, praise and chocolate, please, to:\n%s\n",
                   rnp_get_info("version"),
                   rnp_get_info("maintainer"));
}

/* print a usage message */
void
print_usage(const char *usagemsg)
{
    print_praise();
    (void) fprintf(stderr, "Usage: %s %s", __progname, usagemsg);
}

/* do a command once for a specified file 'f' */
bool
rnp_cmd(rnp_cfg_t *cfg, rnp_t *rnp, optdefs_t cmd, char *f)
{
    const char *key;
    char *      s;

    switch (cmd) {
    case CMD_LIST_KEYS:
    case CMD_LIST_SIGS:
        return match_keys(cfg, rnp, stdout, f, cmd == CMD_LIST_SIGS);
    case CMD_FIND_KEY:
        if ((key = f) == NULL) {
            key = rnp_cfg_getstr(cfg, CFG_USERID);
        }
        return rnp_find_key(rnp, key);
    case CMD_EXPORT_KEY: {
        pgp_dest_t   dst;
        rnp_result_t ret;

        if ((key = f) == NULL) {
            key = rnp_cfg_getstr(cfg, CFG_USERID);
        }

        if (!key) {
            RNP_LOG("key '%s' not found\n", f);
            return 0;
        }

        s = rnp_export_key(rnp, key, rnp_cfg_getbool(cfg, CFG_SECRET));
        if (!s) {
            return false;
        }

        const char *file = rnp_cfg_getstr(cfg, CFG_OUTFILE);
        bool        force = rnp_cfg_getbool(cfg, CFG_FORCE);
        ret = file ? init_file_dest(&dst, file, force) : init_stdout_dest(&dst);
        if (ret) {
            free(s);
            return false;
        }

        dst_write(&dst, s, strlen(s));
        dst_close(&dst, false);
        free(s);
        return true;
    }
    case CMD_IMPORT_KEY:
        if (f == NULL) {
            (void) fprintf(stderr, "import file isn't specified\n");
            return false;
        }
        return rnp_import_key(rnp, f);
    case CMD_GENERATE_KEY:
        key = f ? f : rnp_cfg_getstr(cfg, CFG_USERID);
        rnp_action_keygen_t *        action = &rnp->action.generate_key_ctx;
        rnp_keygen_primary_desc_t *  primary_desc = &action->primary.keygen;
        rnp_key_protection_params_t *protection = &action->primary.protection;
        memset(action, 0, sizeof(*action));
        if (key) {
            strcpy((char *) primary_desc->cert.userid, key);
        }
        primary_desc->crypto.hash_alg = pgp_str_to_hash_alg(rnp_cfg_getstr(cfg, CFG_HASH));
        primary_desc->crypto.rng = &rnp->rng;
        protection->hash_alg = primary_desc->crypto.hash_alg;
        protection->symm_alg = pgp_str_to_cipher(rnp_cfg_getstr(cfg, CFG_CIPHER));
        action->subkey.keygen.crypto.rng = &rnp->rng;

        if (!rnp_cfg_getbool(cfg, CFG_EXPERT)) {
            primary_desc->crypto.key_alg = PGP_PKA_RSA;
            primary_desc->crypto.rsa.modulus_bit_len = rnp_cfg_getint(cfg, CFG_NUMBITS);
            // copy keygen crypto and protection from primary to subkey
            action->subkey.keygen.crypto = primary_desc->crypto;
            action->subkey.protection = *protection;
        } else if (rnp_generate_key_expert_mode(rnp) != RNP_SUCCESS) {
            RNP_LOG("Critical error: Key generation failed");
            return false;
        }
        return rnp_generate_key(rnp);
    case CMD_GET_KEY: {
        char *keydesc = rnp_get_key(rnp, f, rnp_cfg_getstr(cfg, CFG_KEYFORMAT));
        if (keydesc) {
            printf("%s", keydesc);
            free(keydesc);
            return true;
        }
        (void) fprintf(stderr, "key '%s' not found\n", f);
        return false;
    }
    case CMD_TRUSTED_KEYS:
        return rnp_match_pubkeys(rnp, f, stdout);
    case CMD_HELP:
    default:
        print_usage(usage);
        return false;
    }
}

/* set the option */
int
setoption(rnp_cfg_t *cfg, optdefs_t *cmd, int val, char *arg)
{
    switch (val) {
    case OPT_COREDUMPS:
        rnp_cfg_setbool(cfg, CFG_COREDUMPS, true);
        break;
    case CMD_GENERATE_KEY:
        rnp_cfg_setbool(cfg, CFG_NEEDSSECKEY, true);
        *cmd = val;
        break;
    case OPT_EXPERT:
        rnp_cfg_setbool(cfg, CFG_EXPERT, true);
        break;
    case CMD_LIST_KEYS:
    case CMD_LIST_SIGS:
    case CMD_FIND_KEY:
    case CMD_EXPORT_KEY:
    case CMD_IMPORT_KEY:
    case CMD_GET_KEY:
    case CMD_TRUSTED_KEYS:
    case CMD_HELP:
        *cmd = val;
        break;
    case CMD_VERSION:
        print_praise();
        exit(EXIT_SUCCESS);
    /* options */
    case OPT_SSHKEYS:
        rnp_cfg_setstr(cfg, CFG_KEYSTOREFMT, RNP_KEYSTORE_SSH);
        break;
    case OPT_KEYRING:
        if (arg == NULL) {
            (void) fprintf(stderr, "No keyring argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_KEYRING, arg);
        break;
    case OPT_KEY_STORE_FORMAT:
        if (arg == NULL) {
            (void) fprintf(stderr, "No keyring format argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_KEYSTOREFMT, arg);
        break;
    case OPT_USERID:
        if (arg == NULL) {
            (void) fprintf(stderr, "no userid argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_USERID, arg);
        break;
    case OPT_VERBOSE:
        rnp_cfg_setint(cfg, CFG_VERBOSE, rnp_cfg_getint(cfg, CFG_VERBOSE) + 1);
        break;
    case OPT_HOMEDIR:
        if (arg == NULL) {
            (void) fprintf(stderr, "no home directory argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_HOMEDIR, arg);
        break;
    case OPT_NUMBITS:
        if (arg == NULL) {
            (void) fprintf(stderr, "no number of bits argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setint(cfg, CFG_NUMBITS, atoi(arg));
        break;
    case OPT_HASH_ALG:
        if (arg == NULL) {
            (void) fprintf(stderr, "No hash algorithm argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_HASH, arg);
        break;
    case OPT_PASSWDFD:
        if (arg == NULL) {
            (void) fprintf(stderr, "no pass-fd argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_PASSFD, arg);
        break;
    case OPT_RESULTS:
        if (arg == NULL) {
            (void) fprintf(stderr, "No output filename argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_IO_RESS, arg);
        break;
    case OPT_SSHKEYFILE:
        rnp_cfg_setstr(cfg, CFG_KEYSTOREFMT, RNP_KEYSTORE_SSH);
        rnp_cfg_setstr(cfg, CFG_SSHKEYFILE, arg);
        break;
    case OPT_FORMAT:
        rnp_cfg_setstr(cfg, CFG_KEYFORMAT, arg);
        break;
    case OPT_CIPHER:
        rnp_cfg_setstr(cfg, CFG_CIPHER, arg);
        break;
    case OPT_DEBUG:
        rnp_set_debug(arg);
        break;
    case OPT_OUTPUT:
        if (arg == NULL) {
            (void) fprintf(stderr, "No output filename argument provided\n");
            exit(EXIT_ERROR);
        }
        rnp_cfg_setstr(cfg, CFG_OUTFILE, arg);
        break;
    case OPT_FORCE:
        rnp_cfg_setbool(cfg, CFG_FORCE, true);
        break;
    case OPT_SECRET:
        rnp_cfg_setbool(cfg, CFG_SECRET, true);
        break;
    default:
        *cmd = CMD_HELP;
        break;
    }
    return true;
}

/* we have -o option=value -- parse, and process */
int
parse_option(rnp_cfg_t *cfg, optdefs_t *cmd, const char *s)
{
    static regex_t opt;
    struct option *op;
    static int     compiled;
    regmatch_t     matches[10];
    char           option[128];
    char           value[128];

    if (!compiled) {
        compiled = 1;
        if (regcomp(&opt, "([^=]{1,128})(=(.*))?", REG_EXTENDED) != 0) {
            fprintf(stderr, "Can't compile regex\n");
            return 0;
        }
    }
    if (regexec(&opt, s, 10, matches, 0) == 0) {
        (void) snprintf(option,
                        sizeof(option),
                        "%.*s",
                        (int) (matches[1].rm_eo - matches[1].rm_so),
                        &s[matches[1].rm_so]);
        if (matches[2].rm_so > 0) {
            (void) snprintf(value,
                            sizeof(value),
                            "%.*s",
                            (int) (matches[3].rm_eo - matches[3].rm_so),
                            &s[matches[3].rm_so]);
        } else {
            value[0] = 0x0;
        }
        for (op = options; op->name; op++) {
            if (strcmp(op->name, option) == 0) {
                return setoption(cfg, cmd, op->val, value);
            }
        }
    }
    return 0;
}

bool
rnpkeys_init(rnp_cfg_t *cfg, rnp_t *rnp, const rnp_cfg_t *override_cfg, bool is_generate_key)
{
    bool         ret = true;
    rnp_params_t rnp_params;

    rnp_params_init(&rnp_params);
    rnp_cfg_init(cfg);

    rnp_cfg_load_defaults(cfg);
    rnp_cfg_setint(cfg, CFG_NUMBITS, DEFAULT_RSA_NUMBITS);
    rnp_cfg_setstr(cfg, CFG_IO_RESS, "<stdout>");
    rnp_cfg_setstr(cfg, CFG_KEYFORMAT, "human");
    rnp_cfg_copy(cfg, override_cfg);

    memset(rnp, '\0', sizeof(rnp_t));

    if (!rnp_cfg_apply(cfg, &rnp_params)) {
        fputs("fatal: cannot apply configuration\n", stderr);
        ret = false;
        goto end;
    }

    if (rnp_init(rnp, &rnp_params) != RNP_SUCCESS) {
        fputs("fatal: failed to initialize rnpkeys\n", stderr);
        ret = false;
        goto end;
    }

    if (!rnp_key_store_load_keys(rnp, 1) && !is_generate_key) {
        /* Keys mightn't loaded if this is a key generation step. */
        fputs("fatal: failed to load keys\n", stderr);
        ret = false;
        goto end;
    }

end:
    rnp_params_free(&rnp_params);
    if (!ret) {
        rnp_cfg_free(cfg);
        rnp_end(rnp);
    }
    return ret;
}
